src/main/scala/piconot/internal/Language.scala

```diff
@@ -0,0 +1,269 @@
+package piconot.internal
+
+import scala.language.implicitConversions
+import picolib.maze.Maze
+import scala.collection._
+import scalafx.application.JFXApp
+import util.control.Breaks._
+import picolib.semantics._
+import java.io.File
+
+
+
+case class Program(labels: List[Label]) {
+  val processor = new PicoPrgrm("empty.txt", NORTH)
+  System.out.println(processor.labelStates)
+  System.out.println(processor.direction)
+  processor.process(labels, processor.labelStates) // create initial label map
+  val rules = processor.buildrules(labels) // for each element in a label, create a rule
+  processor.execute(rules)
+}
+
+class PicoPrgrm(val mazeFilename: String, dir: Direction) extends JFXApp {
+
+  var labelStates = mutable.Map.empty[String, Int]
+  var direction: Direction = dir
+  var stCounter = 0
+  
+  def execute(rules: List[Rule]) = {
+    val maze = Maze("resources" + File.separator + mazeFilename)
+    object EmptyBot extends Picobot(maze, rules)
+      with TextDisplay with GUIDisplay
+
+    stage = EmptyBot.mainStage
+
+    EmptyBot.run()
+  }
+
+
+  def process(labels: List[Label], labelMap: Map[String, Int]) = {
+    var counter = 1
+
+    System.out.println(labels.toString())
+    // Create a map that goes from top-level label to it's starting state
+    labels.foreach((label: Label) => {
+//      breakable {
+      System.out.println("Label States in first" + labelMap)
+        if (labelMap == Map.empty) {
+         labelMap + (label.name -> 0)
+          System.out.println("Label States in first, if" + labelMap)
+          //break
+        }
+//      }
+      label.rules.foreach((rule: Rules) => {
+        System.out.println("Label States in second" + labelMap)
+        rule match {
+          case Face(dir) => counter = counter + 1
+          case _         => 0
+        }
+      })
+      
+      System.out.println(label)
+      System.out.println(label.name)
+      System.out.println(labelMap)
+      System.out.println(counter)
+      labelMap + (label.name -> counter)
+
+    })
+  }
+
+  def buildrules(rules: List[Label]) = {
+    var builtRules: List[picolib.semantics.Rule] = List()
+
+    rules.foreach((label: Label) => {
+
+      label.rules.foreach((rule: Rules) => {
+        rule match {
+          case Face(dir) => builtRules = builtRules ++ List(evalFace(dir))
+          case Go(sur)   => builtRules = builtRules ++ List(evalGo(sur))
+          case Jump(lbl) => builtRules = builtRules ++ List(evalJump(lbl))
+        }
+
+      })
+    })
+    builtRules
+  }
+
+  def evalFace(dir: Direction): Rule = {
+    direction = dir
+    stCounter += 1
+    Rule(State(stCounter - 1 + ""), Surroundings(Anything, Anything, Anything, Anything), StayHere, State(stCounter + ""))
+  }
+
+  def evalGo(sur: Surrounding): Rule = {
+    var surRule = Surroundings(Anything, Anything, Anything, Anything)
+    var dirRule: MoveDirection = North
+    if (direction == NORTH) {
```

Comment:
This is painful to read. I think a map like `Map<MoveDirection, Map<Surrounding, Surroundings>>` would be cleaner.